# Artificial Neural Networks using Tensorflow

Here, we will be using Tensorflow and keras to build our neural network. Keras is now used as a part of tensorflow (could be used as a standalone package before, but this is better).

Here, we will be working with the MNIST dataset that contains 10's of thousands of black and white images of handwritten numbers. We will be training our model to predict what the written number is.

Note: Don't try to memorize the syntax and functions. Make sure you understand what is going on though.
You will learn the syntax as you practice more problems. [This](https://www.tensorflow.org/api_docs) is the link to TensorFlow documentation. Be careful about the versions though, TF2.0 is fairly different from this version TF 1.13

In [1]:
import tensorflow as tf
import tensorflow.keras as keras
import numpy as np


In [2]:
print(tf.__version__)

1.13.1


### Dataset

This dataset is probably the most common dataset used, and hence, it comes preloaded in keras. As we can see, it is split into images and labels, and further into train and test. 

The labels contain the actual number that is written in the image.

In the following steps, we see that there are 60000 training images, each of the size 28 pixels (height) and 28 pixels (width)

In [3]:
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11493376/11490434 [==============================] - 98s 9us/step


In [4]:
train_images.shape

(60000, 28, 28)

In [5]:
len(train_images)

60000

#### Preprocessing

It is recommended that you normalize your input before sending it to your neural network. A short answer on why from [this](https://www.quora.com/Why-does-mean-normalization-help-in-gradient-descent) Quora post:

"The problem is when you have features with different scales. A commonly used optimization technique, such as gradient descent, uses the product of the learning rate times the gradient of the cost function as the step size. As a result, when the features have different scales, the step size can be different for each feature. When you try different learning rates, you will find that 1) the learning rate is too small and it will take a long time to converge, or 2) the learning rate is too big for one or more features, and it never converges."


In images, the highest pixel value is 255, and lowest is 0. This is a black and white image (grayscale) so images have only one channel (as opposed to Red, Blue, Green channels (3) in full color images).

To scale the values to some value between 0 and 1, one method is to divide every pixel value by 255. And this is what we do in the next step.

In [6]:
train_images = train_images / 255
test_images = test_images / 255

## Building the neural network

We will build a neural network with one input layer, one hidden layer, and one output layer. 
First we flatten the image so we get a vector and not a 2 dimensional image matrix. Here, we need to have 28 x 28 input neurons. Then we have a hidden layer with 128 neurons (it could be any number, there is no real way to tell which is the right size of neurons). And then we have 10 output neurons (because we have ten possible digits (0 - 9)) with the softmax activation. It gives us the number that the image is most likely to be. 

More information is provided in a link at the end of this notebook.

In [7]:
model = keras.Sequential ([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = tf.nn.relu),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

Instructions for updating:
Colocations handled automatically by placer.


We will be using the ```ADAM``` optimizer which another optimizer like gradient descent with a ```sparse_categorical_crossentropy``` loss (this is a solid choice because we are dealing with categorical data).  (you can read more about these at the link provided at the end of the notebook)

In [8]:
model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy',
             metrics = ['accuracy'])

## Training the model

We use ```model.fit(input_data, target_labels)``` to train our model on input data to fit the target labels. We are going to train for five epoch. Each epoch signifies one iteration over the whole training data.

In [9]:
model.fit(train_images, train_labels, epochs = 5)

Epoch 1/5
60000/60000 [==============================] - 12s 206us/sample - loss: 0.2562 - acc: 0.9265
Epoch 2/5
60000/60000 [==============================] - 11s 189us/sample - loss: 0.1119 - acc: 0.9666
Epoch 3/5
60000/60000 [==============================] - 12s 192us/sample - loss: 0.0765 - acc: 0.9770
Epoch 4/5
60000/60000 [==============================] - 12s 202us/sample - loss: 0.0579 - acc: 0.9823
Epoch 5/5
60000/60000 [==============================] - 11s 186us/sample - loss: 0.0446 - acc: 0.9865


### Testing our model

We test our model on the test images and measure its performance by comparing its predictions with the actual test labels.
We see that it gives 97.7% accuracy.

In [10]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(test_acc)

10000/10000 [==============================] - 1s 85us/sample - loss: 0.0743 - acc: 0.9770
0.977


#### More information

go here to see a more detailed explanation on a similiar architecture but on a different dataset. https://www.tensorflow.org/tutorials/keras/basic_classification